### ISSUE TITLE
OpenACC or acceleration on GPU

### ISSUE THREAD

**Issue by echoi** [GeoFLAC/DynEarthSol#9]

**Comment by chaseshyu** [comment]
# Progress Summary

- Pull Request#25 and Reviews [ GeoFLAC/DynEarthSol/pull/25]
- Improvement: Add time information to the replenished marker.
- New test: 250x150x125 km (radius = 2 km) 10 Myr
	- 1.5 GB per frame or .vtu+.vtp files. Elem# = 3M. 
	- Nearly impossible to download and visualize on a PC.

# Remaining decisions

- Switch output format from netCDF4 to HDF5 for directly reading of ParaView?

**Comment by echoi** [comment]
With all the accepted and merged PRs related to this issue, I think this can be closed now.
@chaseshyu Please go ahead and do so if you agree.

**Comment by chaseshyu** [comment]
Actually, there are two unpushed feature updates related to this issue: 1. Memory structure (Structure of Array, SoA) of array2D for better memory access of GPU and 2. knn-bvh (https://github.com/chaseshyu/knn-bvh) Search nearest neighbors of irregular mesh with GPU. The benchmarks in namuscripts are performed with the updated code. There are at least 2-3 PR needed to close this issue. I will push the feature branches once the current RSF feature is merged.
> With all the accepted and merged PRs related to this issue, I think this can be closed now. [@chaseshyu](https://github.com/chaseshyu) Please go ahead and do so if you agree.

**Comment by chaseshyu** [comment]
@echoi @tan2 Another issue I am thinking is if we should keep this new knn-bvh lib repo  (https://github.com/chaseshyu/knn-bvh) in this FLAC organization or just in my personal github? I think this knn-bvh lib should come with DES and also believe this knn-bvh lib might benefit other repo.

> Actually, there are two unpushed feature updates related to this issue: 1. Memory structure (Structure of Array, SoA) of array2D for better memory access of GPU and 2. knn-bvh (https://github.com/chaseshyu/knn-bvh) Search nearest neighbors of irregular mesh with GPU. The benchmarks in namuscripts are performed with the updated code. There are at least 2-3 PR needed to close this issue. I will push the feature branches once the current RSF feature is merged.
> 
> > With all the accepted and merged PRs related to this issue, I think this can be closed now. [@chaseshyu](https://github.com/chaseshyu) Please go ahead and do so if you agree.

**Comment by echoi** [comment]
@chaseshyu I think it's a good idea to bring knn-bvh under GeoFLAC org.

I don't know what's the best practice to distribute knn-bvh with DES. I don't think we would want to make it a static directory under DynEarthSol. Also, I'd like a more convenient way than simply stating in README that  it as a prerequisite.

**Comment by chaseshyu** [comment]
@echoi I think we also can try `git submodule` function, linking a certain version of github repos to current repo as a submodule. For example, the knn-bvh repo uses lbvh as submodule. During the make preceduce, the `git submodule update` will automatic fetch linked lbvh lib.
> [@chaseshyu](https://github.com/chaseshyu) I think it's a good idea to bring knn-bvh under GeoFLAC org.
> 
> I don't know what's the best practice to distribute knn-bvh with DES. I don't think we would want to make it a static directory under DynEarthSol. Also, I'd like a more convenient way than simply stating in README that it as a prerequisite.

**Comment by echoi** [comment]
The submodule approach sounds good to me.

**Comment by tan2** [comment]
Using submodule is fine to me.

I've heard that using submodule requires a specific development workflow.
Can you add some instructions in DEVELOPING?

**Comment by chaseshyu** [comment]
I did not notice DEVELOPING until now ...XD
It seems like the last update was made over a decade ago, in 2012.

> Using submodule is fine to me.
> 
> I've heard that using submodule requires a specific development workflow.
> Can you add some instructions in DEVELOPING?
> […](#)

**Comment by echoi** [comment]
Let's clarify how this approach works and what developers should be mindful of. 
This is important because it could allow other dependencies (e.g., mmg, gospl, and gospl_extensions) to be built automatically as part of the DES build process.

**Comment by chaseshyu** [comment]
The instruction has been updated here in `DEVELOPING.md`

https://github.com/GeoFLAC/DynEarthSol/blob/62a0e4f4415ba56ce91f5e666042247c9a2ea7db/DEVELOPING.md?plain=1#L38

> Using submodule is fine to me.
> 
> I've heard that using submodule requires a specific development workflow.
> Can you add some instructions in DEVELOPING?
> […](#)
